File: 019_remove_nth_node_from_end_of_list/Solution1.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
    }
}

public class Solution1 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
       return removeNode(head, n) == n ? head.next : head; 
    }

    public int removeNode(ListNode node, int n) {
        if (node.next == null) return 1;
        int m = removeNode(node.next, n);
        if (m == n) {
            node.next = node.next.next;
        }

        return m + 1;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);

        Solution1 solution = new Solution1();
        ListNode result = solution.removeNthFromEnd(head, 2);
        while (result != null) {
            System.out.print(result.val + " ");
            result = result.next;
        }
    }
}

File: 019_remove_nth_node_from_end_of_list/Solution1.go
package main

import "fmt"

type ListNode struct {
    Val  int
    Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
    if removeNode(head, n) == n {
        return head.Next
    } else {
        return head
    }
}

func removeNode(head *ListNode, n int) int {
    if head.Next == nil {
        return 1
    }

    m := removeNode(head.Next, n)
    if m == n {
        head.Next = head.Next.Next
    }

    return m + 1
}

func main() {
	head := &ListNode{1, nil}
	head.Next = &ListNode{2, nil}
	head.Next.Next = &ListNode{3, nil}
	head.Next.Next.Next = &ListNode{4, nil}
	head.Next.Next.Next.Next = &ListNode{5, nil}

	result := removeNthFromEnd(head, 2)
	for result != nil {
		fmt.Printf("%d ", result.Val)
		result = result.Next
	}
}

